ext/crates/fp/src/matrix/subspace.rs

```diff
@@ -298,3 +298,50 @@ impl std::fmt::Display for Subspace {
         Ok(())
     }
 }
+
+#[cfg(feature = "proptest")]
+pub mod arbitrary {
+    use proptest::prelude::*;
+
+    use super::*;
+    use crate::matrix::matrix_inner::arbitrary::MatrixArbParams;
+    pub use crate::matrix::matrix_inner::arbitrary::MAX_COLUMNS as MAX_DIM;
+
+    #[derive(Debug, Clone)]
+    pub struct SubspaceArbParams {
+        pub p: Option<ValidPrime>,
+        pub dim: BoxedStrategy<usize>,
+    }
+
+    impl Default for SubspaceArbParams {
+        fn default() -> Self {
+            Self {
+                p: None,
+                dim: (0..=MAX_DIM).boxed(),
+            }
+        }
+    }
+
+    impl Arbitrary for Subspace {
+        type Parameters = SubspaceArbParams;
+        type Strategy = BoxedStrategy<Self>;
+
+        fn arbitrary_with(args: Self::Parameters) -> Self::Strategy {
+            let p = match args.p {
+                Some(p) => Just(p).boxed(),
+                None => any::<ValidPrime>().boxed(),
+            };
+
+            (p, args.dim)
+                .prop_flat_map(move |(p, dim)| {
+                    Matrix::arbitrary_with(MatrixArbParams {
+                        p: Some(p),
+                        rows: (0..=dim + 1).boxed(),
+                        columns: Just(dim).boxed(),
+                    })
```

Comment:
Is this likely to generate interesting subspaces? Almost all will be of full or nearly full rank. I'd think you should first pick the rank k, generate an arbitrary rank k * dim matrix, and then for each additional row pick a random linear combination of the first k rows. The last row can be left as all 0. Alternatively, it might be better to first pick a random row echelon form and then randomly pick the entries in the non-pivot columns.

Reply:
The matrices themselves will be full rank the majority of the time, but they'll define a subspace of an ambient space of fixed dimension `dim`. The dimension of the subspace itself is the rank of the matrix, but because the number of rows ranges over `0..=dim+1`, I think we hit every subspace that way.

Is your point that we should test creating subspaces with highly rank deficient matrices? One of the invariants of `Subspace` is that it always contains a row-reduced matrix, so it'll be the same as creating a subspace with a smaller matrix and extending with rows of 0s. We could add a test that creates those deficient matrices and checks that the matrix does become row-reduced

Reply:
That's good at least, but won't we still almost always end up in the dense Schubert cell for `Gr_{k, dim}` where the pivots lie along the main diagonal? I'd think what we'd want is to pick a random subset of `k` of the `dim` columns to be the pivot columns, and then fill the non pivot columns with random data.

Reply:
I simulated that by building an arbitrary matrix and forcing some pivots by filling initial segments with 0s. That should give more interesting subspaces